doc/rfc008-managing-apalache.md

```diff
@@ -0,0 +1,343 @@
+The crux here is that both an upgrade and a downgrade of Apalache, relative to a
+specific Quint release, may maintain or break compatibility.
+
+Given that Quint calls Apalache, but not the other way around, it makes sense to
+maintain a list of compatible Apalache versions for a given Quint release.
+Broadly speaking, there's two options for maintaining such a list:
+
+1. Informally, as part of the release notes, or other documentation.
+2. Formally, inside the Quint codebase.
+
+Moreover, there's multiple options on how to specify compatible versions:
+
+#### 3.3.1 Use arbitrary Apalache release
+
+We don't expect a specific Apalache release from Quint.
+
+Pros:
+
+- No need to manage or communicate version dependencies.
+
+Cons:
+
+- Things break at runtime with incompatible versions, possibly in weird places
+  and with strange error messages or uncaught exceptions.
+
+#### 3.3.2 Use latest Apalache release
+
+Regardless of the Quint release, we expect the latest version of Apalache to be
+installed. We could use the Github REST API to check the latest Apalache release
+and abort if the installed version differs.
+
+Pros:
+
+- Easy management, no need to pin specific versions from Quint.
+
+Cons:
+
+- Prone to break if users have an outdated version of Quint.
+  - We could address this by also checking the latest release of Quint, and
+    alerting users to upgrade.
+- Prone to break if users downgrade their installation of Quint.
+
+#### 3.3.3 Pin a specific Apalache version
+
+Each Quint release pins a specific version of Apalache.
+
+Pros:
+
+- Enables user to downgrade Quint, while maintaining compatibility with Apalache.
+
+Cons:
+
+- Users don't receive Apalache bugfix releases, unless we cut another Quint
+  release that pins the updated dependency.
+
+#### 3.3.4 Pin a minor release branch
+
+Each Quint release pins an Apalache minor release, but allows patch-level
+updates. This would correspond to a [tilde range](https://docs.npmjs.com/cli/v6/using-npm/semver#tilde-ranges-123-12-1)
+`~major.minor.patch` in common package managers.
+
+Pros:
+
+- Users receive patch versions of Apalache while maintaining compatibility.
+- Enables user to downgrade Quint, while maintaining compatibility with Apalache.
+
+Cons:
+
+- There may be an Apalache minor version bump that does not break compatibility
+  Quint. We would need to cut another Quint release that pins the updated minor
+  release.
+
+## 4. Proposed Solution
+
+We choose managed options for obtaining and launching Apalache from Quint, while
+still providing a non-managed option:
+
+1. Primarily, Quint uses a Quint-managed, on-demand instance of Shai:
+   1. If not present, Quint fetches a compatible Apalache release from Github,
+      using [Github's REST API endpoint][] and/or [`octokit/request.js`][] and
+      unpacks it into a local installation directory.
+      - This can be scripted etirely from NodeJS, foregoing the need to wrap a
+        third-party package manager binary.
```

Comment:
If we really think it is so important to dynamically determine the latest apalche version, then I think we should not run this by default on every invocation. Rather, lets protect it behind an update flag and/or gate any automatic checks via a periodic (e.g., once per week) check.

We should also, ideally, have a way to fall back to a known good version url, i.e., the minimal minor version `m.n.0`, so that our CLI doesn't break due to api limits.

That all said, I think it is probably fine for now to just hard-code the compatible apalache version and push updates as needed via releases. If that gets to be a bottleneck, then we can optimize. This seems a bit premature to me, in light of the complexity costs we see in, e.g., #1124.

Reply:
"this" is meant to refer to the design of dynamically determining and updating to the latest patch version of apalache. This introduces a lot of complexity and variable dependencies afaict, 'cause we are ending up trying to implement like a fragment of a package manager :)

I actually disagree that my suggestion increases complexity of the code base, or the UI. Afaiu, it would only be a slight increase in the *amount* of work for releases, but actually would simplify rather than complicate our dependency on Apalache. My perspective is tied to the view that complexity is about how difficult it is to understand or reason about a system. As a user, I'd rather know when things are going to update and control that rather than have it automatically and without my knowledge performing updates every time I run it. The former is *simpler*, in my view, even tho it requires me taking an extra action when I want to update,  because it is easier to understand what is going on and how the system works.

That said, I do think there are ways I think we can support this variable version approach without hitting a bottleneck on the github API. I just can't think of any that don't require us to invest in developing a bigger fragment of a package manager, and I just don't see how that lies on our critical path atm! :sweat_smile: 

But maybe this is best for a sync discussion next week! Nothing urgent to address I think!

Reply:
> As a user, I'd rather know when things are going to update and control that rather than have it automatically and without my knowledge performing updates every time I run it.

I think your concern is taken care of under §3 below:
> To give the user more control over their system, or as a development setup, we provide a CLI flag and/or config file option [...] Quint skips the managed approach above and instead connects to the provided socket and uses this instance of Apalache dispatch verification commands.

You are free to maintain that version of Apalache in any way you like.
It would be helpful to understand if this covers your needs, and, if not, why?

> we are ending up trying to implement like a fragment of a package manager

> I'd rather know when things are going to update and control that rather than have it automatically and without my knowledge performing updates every time I run it

> As a user, I'd rather know when things are going to update and control that rather than have it automatically and without my knowledge performing updates every time I run it.

I think this is a biased framing of the situation, and I would prefer to stick to the facts in a technical discussion. You do know when things are going to update (when you run `verify`), it's not automatic but user-initiated, and it's printing messages so nothing goes on without your knowledge.

And no, I'm definitely not trying to implement a package manager.

Reply:
That said, @ivan-gavran mentioned yesterday that the MBT team hit similar questions when developing Atomkraft. I will do some research next week where there discussion ended up, so we don't unnecessarily duplicate efforts.

Reply:
Let’s talk about this sync next week! Sorry if my remarks came across as flippant.

To be clear, I don’t think that the fact this is tending towards the problems of package management is a fault in your efforts or thinking, imo that’s just the kind of problem space we are in. It’s a complex domain where there isn’t a great, widely used, cross platform solution, and figuring out a simple but flexible solution to our needs is a significant design task imo.